### PR TITLE
Refactoring taint functions to reduce sprawl

### LINKED ISSUE #45060: Refactoring taints logic to reduce sprawl across components.
As a follow-up on the discussion that we had on #43171 and sig-scheduling meeting, I wish to consolidate the points that were raised there wrt taints logic sprawl in kubectl.

- Some functions like reorganizeTaints in taints.go in pkg/kubectl/cmd/ need to moved to another location(either api/helpers.go or new TaintController pkg) - @gmarek, @timothysc 
- Refactor the testing suite to reflect the new changes.

Those are specific to the kubectl. 

/cc @timothysc @gmarek @aveshagarwal @k82cn @adohe @jayunit100

Please add other files if you happen(ned) to work or if you know.

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #45060

**Special notes for your reviewer**:
@gmarek @timothysc @k82cn @jayunit100 - I moved some fn's to helpers and some to utils. LMK, if you are ok with this change.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by cblecker** [comment]
@ravisantoshgudimetla PR needs rebase

**Comment by gyliu513** [comment]
ping @ravisantoshgudimetla any update for this?

**Comment by ravisantoshgudimetla** [comment]
@gyliu513 I am holding off the rebase till 1.7 release is out, so that I don't have to rebase it again later but if you think that this could be part of 1.7, I can rebase this PR now.

**Comment by gyliu513** [comment]
make sense, thanks @ravisantoshgudimetla , let's wait this after 1.7 release.

**Comment by ravisantoshgudimetla** [comment]
/sig scheduling

Rebased the files. This is up for review now.

File: pkg/api/v1/helper/helpers.go [hunk 1]
```diff
@@ -513,3 +519,70 @@ func StorageNodeAffinityToAlphaAnnotation(annotations map[string]string, affinit
 	annotations[v1.AlphaStorageNodeAffinityAnnotation] = string(json)
 	return nil
 }
+
+// ReorganizeTaints returns the updated set of taints, taking into account old taints that were not updated,
+// old taints that were updated, old taints that were deleted, and new taints.
+func ReorganizeTaints(node *v1.Node, overwrite bool, taintsToAdd []v1.Taint, taintsToRemove []v1.Taint) (string, []v1.Taint, error) {
```
Comment (gyliu513) [inline pkg/api/v1/helper/helpers.go]: @ravisantoshgudimetla Can you please show more detail for what is the intention of putting those functions such as `ReorganizeTaints`, `addTaints`, `deleteTaints` and `ValidateNoTaintOverwrites` to the `helper` folder but not the `taintutil` like `ParseTaints`?

Is there are policy for this such as when to update `helper` and when to update `util`?

Comment (ravisantoshgudimetla) [inline pkg/api/v1/helper/helpers.go]: > Is there are policy for this such as when to update helper and when to update **util?**

I am not aware of anything like that.


So, there is no consensus on where these fns have to be put and both places seemed reasonable. Please look at https://github.com/kubernetes/kubernetes/issues/45060#issuecomment-299125377

The following are the reasons that I chose those respective places.
- /pkg/util/taints/taints.go already had the fns that are needed for taint operations like Set, string conversion etc which are kind of basic operations on taints(I saw them more like primitive operations. For ex, all the fn's here operate on string, []string etc).
- /pkg/v1/helper operates only on the taints on node and fn's in this package may need some fn's from util package.

Again, these are my views and I am fine with making changes.

File: pkg/api/v1/helper/helpers.go [hunk 2]
```diff
@@ -513,3 +519,70 @@ func StorageNodeAffinityToAlphaAnnotation(annotations map[string]string, affinit
...
+	newTaints := append([]v1.Taint{}, taintsToAdd...)
+	oldTaints := node.Spec.Taints
+	// add taints that already existing but not updated to newTaints
+	added := addTaints(oldTaints, &newTaints)
```
Comment (k82cn) [inline pkg/api/v1/helper/helpers.go]: why using list/slice pointer?

File: pkg/api/v1/helper/helpers.go (hunk 1 above)
Comment (gyliu513) [inline pkg/api/v1/helper/helpers.go]: Thanks @ravisantoshgudimetla for the explain, I was thinking put all of the functions to `taint util` maybe better as I can get all of the util fns for taint in one place and do not need to switch to other places. 

Let us wait what other's point for this issue.

File: pkg/api/v1/helper/helpers.go (hunk 2 above)
Comment (gyliu513) [inline pkg/api/v1/helper/helpers.go]: This is just a refactor and do not touch the old logic, if needed, I think that this should be handled in a separated PR.

File: pkg/api/v1/helper/helpers.go [hunk 3]
```diff
@@ -513,3 +519,70 @@ func StorageNodeAffinityToAlphaAnnotation(annotations map[string]string, affinit
...
+	allErrs, deleted := deleteTaints(taintsToRemove, &newTaints)
```
Comment (k82cn) [inline pkg/api/v1/helper/helpers.go]: will `taintsToAdd` and `taintsToRemove` have the same item? what's expected behaviour?

File: pkg/api/v1/helper/helpers.go [hunk 4]
```diff
@@ -513,3 +519,70 @@ func StorageNodeAffinityToAlphaAnnotation(annotations map[string]string, affinit
...
+		}
+		if !existsInNew {
+			*newTaints = append(*newTaints, oldTaint)
+		}
+	}
+	return len(oldTaints) != len(*newTaints)
+}
+
+// ValidateNoTaintOverwrites validates that when overwrite is false, to-be-updated taints don't exist in the node taint list (yet)
+func ValidateNoTaintOverwrites(node *v1.Node, taints []v1.Taint) error {
```
Comment (k82cn) [inline pkg/api/v1/helper/helpers.go]: this func is too specific to cli's requirement, prefer something like `Intersection`.

File: pkg/api/v1/helper/helpers.go [hunk 5]
```diff
@@ -23,12 +23,18 @@ import (
...
 	"k8s.io/apimachinery/pkg/selection"
-
 	"k8s.io/api/core/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/api/helper"
 )
 
+const (
+	MODIFIED  = "modified"
```
Comment (k82cn) [inline pkg/api/v1/helper/helpers.go]: Those const seems only used by cli; can we just keep them in cli?

**Comment by k82cn** [comment]
@ravisantoshgudimetla , comments overall:
1. some func are specific to cli's logic, prefer to make them more general, e.g.  `ValidateNoTaintOverwrites` can change to check intersection of node.Taints and newTains, if intersection is not 0 & overwrite is true, report error.
2. some func in v1helper and some func in util.taint, any guidance for that
3. related to # 2, it seems we did not include `pkg/util` in client-go; if using client-go in future, do we need another PR for that? cc/ @caesarxuchao for input

File: pkg/api/v1/helper/helpers.go (hunk 4 above)
Comment (ravisantoshgudimetla) [inline pkg/api/v1/helper/helpers.go]: Makes sense. I will change it.

File: pkg/api/v1/helper/helpers.go (hunk 5 above)
Comment (ravisantoshgudimetla) [inline pkg/api/v1/helper/helpers.go]: These are needed for the reflecting node changes. So, I guess these are needed here.

File: pkg/api/v1/helper/helpers.go (hunk 3 above)
Comment (ravisantoshgudimetla) [inline pkg/api/v1/helper/helpers.go]: They are different. For example when someone wants to delete taints and add taints in single command like
` kubectl taint node1 dedicated1=sample=a dedicated-`

dedicated1 will be taintsToAdd and dedicated will be taintsToRemove.

**Comment by caesarxuchao** [comment]
We talked about creating a k8s.io/util repo. @monopole @pwittrock is k8s.io/util on your radar?

File: pkg/api/v1/helper/helpers.go (hunk 3 above)
Comment (k82cn) [inline pkg/api/v1/helper/helpers.go]: Did not test; but what's the action if we run `kubectl taint node1 dedicated1=sample=a dedicated1-`?

Comment (ravisantoshgudimetla) [inline pkg/api/v1/helper/helpers.go]: We will get an error like this:

error: can not both modify and remove the following taint(s) in the same command: {"sample":""}

**Comment by ravisantoshgudimetla** [comment]
@k82cn I made the 1st change suggested but with small refining. Instead of returning bool, I am returning a string which has taintKeys seperated by `,` , so that we can tell the user what all keys are already existing.

Regarding points 2&3, I will wait for others to comment on.

Friendly ping @timothysc @gmarek @aveshagarwal

**Comment by k82cn** [comment]
@monopole @pwittrock , what's the plan of `k8s.io/util`? when we do some refactor work, e.g. this PR, prefer to align with your plan, e.g. moving all helper funcs in to `pkg/api/helper`; so we can avoid missing as much as possible :).

**Comment by k82cn** [comment]
Regarding 1st comments, honestly, I'd like to have a Taint Set operations, e.g. merge, sub/delete; this kind of helper is more general for other to use.

**Comment by ravisantoshgudimetla** [comment]
> Regarding 1st comments, honestly, I'd like to have a Taint Set operations, e.g. merge, sub/delete; this kind of helper is more general for other to use.

So, if I understood correctly, you are suggesting to have some util package which has set operations on not only taints but for other items as well(like labels etc).

**Comment by k82cn** [comment]
nop, export set operation for taints; NO other items, e.g. label.

File: pkg/api/v1/helper/helpers.go (hunk 5 above)
Comment (timothysc) [inline pkg/api/v1/helper/helpers.go]: I would prefer consts to be in a common location of a scheduling library, not sprinkled into the apimachinery.  This defeats the purpose.

File: pkg/api/v1/helper/helpers.go (hunk 1 above)
Comment (timothysc) [inline pkg/api/v1/helper/helpers.go]: same

**Review by timothysc — CHANGES_REQUESTED** [review]
I'm  against moving common code into api-helpers.  

The code should be in some common utility library either in the scheduler or other location.

**Comment by k82cn** [comment]
@timothysc , I'm thinking 'api/helper' is the 'common utility library' :).

**Comment by timothysc** [comment]
> I'm thinking 'api/helper' is the 'common utility library' 

Please no... api should be touched so sparingly if/when possible.

**Comment by ravisantoshgudimetla** [comment]
> Please no... api should be touched so sparingly if/when possible.

Would pkg/uti/taints.go be a good place? @timothysc

**Comment by timothysc** [comment]
> Would pkg/uti/taints.go be a good place? @timothysc 

Yes please.

**Comment by ravisantoshgudimetla** [comment]
Ok, I will move the fn's to that location. Thnx.

**Comment by ravisantoshgudimetla** [comment]
@timothysc @k82cn @gyliu513 Moved the fn's as suggested.

**Comment by ravisantoshgudimetla** [comment]
Accidentally closed the PR and re-opened it.

File: pkg/util/taints/taints.go [hunk 6]
```diff
@@ -94,3 +103,198 @@ func (t taintsVar) String() string {
...
+		if taintToDelete.MatchTaint(&taints[i]) {
+			deleted = true
+			continue
+		}
+		newTaints = append(newTaints, taints[i])
+	}
+	return newTaints, deleted
+}
+
+// Tries to remove a taint from annotations list. Returns a new copy of updated Node and true if something was updated
```
Comment (k82cn) [inline pkg/util/taints/taints.go]: start with `RemoveTaint`.

File: pkg/util/taints/taints.go [hunk 7]
```diff
@@ -94,3 +103,198 @@ func (t taintsVar) String() string {
...
+	if !TaintExists(nodeTaints, taint) {
+		return newNode, false, nil
+	}
+
+	newTaints, _ := DeleteTaint(nodeTaints, taint)
+	newNode.Spec.Taints = newTaints
+	return newNode, true, nil
+}
+
+// Tries to add a taint to annotations list. Returns a new copy of updated Node and true if something was updated
```
Comment (k82cn) [inline pkg/util/taints/taints.go]: ditto

File: pkg/util/taints/taints.go [hunk 8]
```diff
@@ -94,3 +103,198 @@ func (t taintsVar) String() string {
...
+
+	if !updated {
+		newTaints = append(newTaints, *taint)
+	}
+
+	newNode.Spec.Taints = newTaints
+	return newNode, true, nil
+}
+
+func TaintExists(taints []v1.Taint, taintToFind *v1.Taint) bool {
```
Comment (k82cn) [inline pkg/util/taints/taints.go]: please add comments for it.

**Comment by ravisantoshgudimetla** [comment]
Any other comments?

File: pkg/util/taints/taints.go [hunk 9]
```diff
@@ -94,3 +103,199 @@ func (t taintsVar) String() string {
 func (t taintsVar) Type() string {
 	return "[]api.Taint"
 }
+
+// ParseTaints takes a spec which is an array and creates slices for new taints to be added, taints to be deleted.
+func ParseTaints(spec []string) ([]v1.Taint, []v1.Taint, error) {
```
Comment (timothysc) [inline pkg/util/taints/taints.go]: Are there unit tests for these functions?  If not we should add them.

**Review by timothysc — APPROVED** [review]

File: pkg/util/taints/taints.go (hunk 9 above)
Comment (ravisantoshgudimetla) [inline pkg/util/taints/taints.go]: As of now, none, I want to add them in a separate PR as this is just for moving the fn's.

**Comment by ravisantoshgudimetla** [comment]
Thanks for the review @timothysc. 

/assign @smarterclayton

**Pull request merged**